### PR TITLE
docs: document CATS DSL format

### PR DESCRIPTION
**Pull request by Jaguar0625**

### CODE REVIEW

File: docs/cats_dsl.md [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+import "other.cats"
+```
+
+All imported filenames are relative to the include path passed to the parser.
+Using the CLI tool, this can be specified by the `--include` command line argument.
+
+### using
+
+Alias statements can be used to name a unique POD type.
+CATS supports two types of built-ins.
```
Comment (0x6861746366574) [inline docs/cats_dsl.md:20]: Change to a colon.

File: docs/cats_dsl.md [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+Integer types: unsigned ("uint") and signed ("int") for sizes { 1, 2, 4, 8 }.
```
Comment (0x6861746366574) [inline docs/cats_dsl.md:22]: Should these be bullets (-) instead?

File: docs/cats_dsl.md [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+For example, to define a `Height` type that is represented by a 8 byte unsigned integer:
+```cpp
+using Height = uint64
+```
+
+Fixed buffer types: unsigned byte buffer of any length.
```
Comment (0x6861746366574) [inline docs/cats_dsl.md:28]: Should this be a bullet (-) point instead?

File: docs/cats_dsl.md [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+Dynamically sized arrays are supported.
+Each array has an associated size that can be a constant, a property reference or a special `__FILL__` keyword.
+
+For example, to define a `Garage` with a `vehicles` field that is composed of `vehicles_count` `Vehicle` structures:
+```cpp
+struct Garage
+	vehicles_count = uint32
+
+	vehicles = array(Vehicle, vehicles_count)
```
Comment (gimre-xymcity) [inline docs/cats_dsl.md:125]: in this form it only makes sense if Vehicle is constant sized, no?
that is with this Vehicle needs @size annotation...

i.e. compare with `address_additions = array(UnresolvedAddress, address_additions_count)`

File: docs/cats_dsl.md [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+```
+
+The special `__FILL__` keyword indicates that the array extends until the end of the structure.
+For example, to indicate the `vehicles` array composed of `Vehicle` structures extends to the end of the `Garage` structure:
+```cpp
+struct Garage
+	vehicles = array(Vehicle, __FILL__)
+```
```
Comment (gimre-xymcity) [inline docs/cats_dsl.md:133]: this doesn't mention that in this case Garage needs to have `@size` ...
`__FILL__` cannot work without `@size`, cause _number of elements_ == `(.size - "current offset") / object.size`

also I don't think we currently allow `__FILL__` with variable-sized objects

File: docs/cats_dsl.md [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+```cpp
+struct Car
+	weight = uint32
+
+	max_clearance = Height
+	has_left_steering_wheel = uint8
+```
+
+In addition, a named inline will inline a referenced structure's fields with a prefix.
+For example, in the following `SizePrefixedString` is inlined in `Vehicle` as `friendly_name`:
```
Comment (gimre-xymcity) [inline docs/cats_dsl.md:157]: mention that `__value__` is a special keyword?

File: docs/cats_dsl.md [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+```cpp
+inline struct SizePrefixedString
+	size = uint32
```
Comment (gimre-xymcity) [inline docs/cats_dsl.md:160]: maybe use some other name, so that it's clear that generated name is concatenation of two names

**Review by gimre-xymcity — CHANGES_REQUESTED** [review]

**Review by 0x6861746366574 — APPROVED** [review]
Some minor stylistic changes.

File: docs/cats_dsl.md (hunk 4 above)
Comment (0x6861746366574) [reply docs/cats_dsl.md:125]: Isn't `size` an attribute only for variable-sized structures, not constant?

Comment (Jaguar0625) [reply docs/cats_dsl.md:125]: although we don't have anything like that currently (and generator doesn't support it), i think it correct and makes sense from schema point of view

File: docs/cats_dsl.md (hunk 5 above)
Comment (Jaguar0625) [reply docs/cats_dsl.md:133]: we do
```
	transactions = array(Transaction, __FILL__)
```

Comment (gimre-xymcity) [reply docs/cats_dsl.md:133]: > also I don't think we currently allow __FILL__ with variable-sized objects

actually I see py generator supports that ^^

anyway, what about `@size` on Garage? how is generator supposed to know what variable to modify / what variable to use when reading "fill array"?

**Review by gimre-xymcity — APPROVED** [review]

**Pull request merged**